ga4gh/avrotools.py

```diff
@@ -129,309 +92,419 @@ def _ensureProtocolSubclass(self):
                 self.class_.__name__)
             raise SchemaToolException(message)
 
-    def _ensureNoExtraFields(self, jsonDict):
-        # TODO we could probably move this inside _getInvalidFields
-        # when the underlying object is a python dict
-        schemaJson = self.schema.to_json()
-        schemaJsonFields = schemaJson['fields']
-        fieldNamesSet = set()
-        for field in schemaJsonFields:
-            fieldName = field['name']
-            fieldNamesSet.add(fieldName)
-        for fieldName in jsonDict.keys():
+
+class AvroTypeSwitch(object):
+    """
+    Provides reusable logic for branching based on avro types
+    """
+    schemaRequired = set(
+        ['record', 'enum', 'map', 'array', 'union',
+         'request', 'error', 'error_union', 'fixed'])
+
+    def __init__(self, class_):
+        self.class_ = class_
+        self.schema = class_.schema
+
+    def getInstance(self):
+        jsonDict = self.handleSchema(self.schema)
+        instance = self.class_.fromJsonDict(jsonDict)
+        return instance
+
+    def getFieldValue(self, fieldName):
+        fieldType = None
+        for field in self.schema.fields:
+            if field.name == fieldName:
+                fieldType = field.type
+                break
+        value = self.handleSchema(fieldType)
+        return value
+
+    def handleSchema(self, schema, extra=None):
```

Comment:
Docstring. Since we don't have docstrings for the individual `handleX` methods we should provide a good overview of how this setup works in this function.